(Refactor) Lot in Stock

DedrickEnc opened the pull request:
This PR closes #1848 

In this pull request, the following action was performed : 

- The registry was updated to follow the new standard

- Two components, ```bhDepotSelect``` and ```bhInventorySelect``` were added

- Two tests components ware added into the folder ```/shared/components```

- The integration test was updated 

- The filter parser at the server side (```stock/core.js```) was configured differently, ```autoParseStatements : false```

- The end to end test for lot in stock module uses now the shared modal search page

DedrickEnc committed:
applting new filter

DedrickEnc committed:
add a component to select depot

DedrickEnc committed:
replace old code by component and apply filter at the server side

DedrickEnc committed:
add a period select component

DedrickEnc committed:
research functionnality done

DedrickEnc committed:
lot i stock done

DedrickEnc committed:
fixing integration test

DedrickEnc committed:
update integration and e2e test

jniles commented on client/src/js/components/bhDepotSelect.js:
```diff
@@ -0,0 +1,45 @@
+    Depots.read()
+      .then(function (depots) {        
+        $ctrl.depots = depots;
+      });
```
This should use `Notify` to handle errors.

jniles commented on client/src/js/components/bhDepotSelect.js:
```diff
@@ -0,0 +1,45 @@
+        $ctrl.depots = depots;
+      });
+
+    $ctrl.valid = true;
```
Do you ever use this?  I couldn't find where it might be used.

jniles commented on client/src/js/components/bhInventorySelect.js:
```diff
@@ -0,0 +1,45 @@
+    Inventory.read()
+      .then(function (inventories) { 
+        $ctrl.inventories = inventories;
+      });
```
This also need `Notify` to handle errors.

jniles commented on client/src/js/components/bhInventorySelect.js:
```diff
@@ -0,0 +1,45 @@
+        $ctrl.inventories = inventories;
+      });
+
+    $ctrl.valid = true;
```
Do you use this?

jniles commented on client/src/modules/stock/lots/modals/search.modal.html:
```diff
@@ -11,58 +11,83 @@
-      <label class="control-label" translate>STOCK.DEPOT</label>
+          <!-- Inventaire -->
+          <bh-inventory-select
+            inventory-id="$ctrl.searchQueries.inventory_uuid"
```
Could you rename this parameter `inventory-uuid`?  Since there isn't actually an inventory ID column, it will save confusion in the future.

Note that the `name` here should also be updated.

jniles commented on client/src/modules/stock/lots/modals/search.modal.html:
```diff
@@ -11,58 +11,83 @@
+        <div class="tab-body">
+          <!-- depot -->          
+          <bh-depot-select
+            depot-id="$ctrl.searchQueries.depot_uuid"
```
Could you rename this parameter `depot-uuid`?  It will save confusion in the future.  Note that the `name` property of this component will also need to be updated.

jniles commented on test/end-to-end/stock/stock.lots.spec.js:
```diff
@@ -1,84 +1,84 @@
-    FU.uiSelect('$ctrl.bundle.depot_uuid', 'Depot Principal');
-    FU.modal.submit();
-    GU.expectRowCount(gridId, 8 + depotGroupingRow);
+  it('finds ten lot for all time', () => {
```
This should read "sixteen".  You might try doing this:
```js
it(`finds ${LOT_FOR_ALLTIME} lots for all time`, () => {
  // etc...
});
```

jniles commented on test/end-to-end/stock/stock.lots.spec.js:
```diff
@@ -1,84 +1,84 @@
   const depotGroupingRow = 1;
 
-  it('find lots by depot', () => {
+  it('finds ten lot for today', () => {
```
:+1:

jniles commented on server/controllers/stock/core.js:
```diff
@@ -78,8 +85,9 @@ function getLots(sqlQuery, parameters, finalClauseParameter) {
 
   filters.dateFrom('dateFrom', 'date', 'm');
   filters.dateTo('dateTo', 'date', 'm');
+  filters.setGroup(finalClause || '');
```
Wow, this works but it is confusing.  Could you put a comment above this line telling future programmers that `FilterParser` sets the grouping to an empty string when it doesn't exist?  I had to look it up ([filter.js#L57](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/server/lib/filter.js#L57).)

jniles commented on server/controllers/stock/core.js:
```diff
@@ -78,8 +85,9 @@ function getLots(sqlQuery, parameters, finalClauseParameter) {
+  filters.setGroup(finalClause || '');
 
-  const query = filters.applyQuery(sql).concat(finalClause || '');
+  let query = filters.applyQuery(sql);
```
Does this need to be a `let` now?

jniles commented on client/src/modules/vouchers/vouchers.service.js:
```diff
@@ -146,7 +146,7 @@ function VoucherService(Api, $http, util, TransactionTypeStore, Modal,
   function reverse(creditNote) {
     return $http.post(baseUrl.concat(creditNote.uuid, '/reverse'), creditNote)
-      .then(util.unwrapHttpResponse);
+      .then(service.util.unwrapHttpResponse);
```
:+1:

jniles commented on client/src/modules/templates/bhInventorySelect.tmpl.html:
```diff
@@ -0,0 +1,24 @@
+<div ng-form="{{ $ctrl.name }}" bh-inventory-select ng-model-options="{ updateOn: 'default' }">
+  <div
+    class="form-group"
+    ng-class="{ 'has-error' : InventoryForm.$$parent.$submitted && InventoryForm.inventory_id.$invalid }">
```
Where is `InventoryForm` defined?

I think you can probably remove the `ng-form = "{{ $ctrl.name }}"` call and just use `ng-form="InventoryForm"`.  That should work well and eliminate need for the `name` parameter in the `bhInventorySelect` component :+1:

jniles commented on client/src/modules/templates/bhDepotSelect.tmpl.html:
```diff
@@ -0,0 +1,24 @@
+<div ng-form="{{ $ctrl.name }}" bh-depot-select ng-model-options="{ updateOn: 'default' }">
+  <div
+    class="form-group"
+    ng-class="{ 'has-error' : DepotForm.$$parent.$submitted && DepotForm.depot_id.$invalid }">
```
See my comment on the `bhInventorySelect` - the same logic can be applied here.

jniles commented on client/src/modules/templates/bhDepotSelect.tmpl.html:
```diff
@@ -0,0 +1,24 @@
+      ng-model="$ctrl.depotId"
+      on-select="$ctrl.onSelect($item, $model)">
+      <ui-select-match placeholder="{{ 'FORM.SELECT.DEPOT' | translate }}">{{$select.selected.text}}</span></ui-select-match>
+      <ui-select-choices ui-select-focus-patch repeat="depot.uuid as depot in $ctrl.depots | filter: { 'text': $select.search }">
```
Good use of filter :+1:

jniles commented on client/src/modules/templates/bhDepotSelect.tmpl.html:
```diff
@@ -0,0 +1,24 @@
+    <ui-select name="depot_id" 
+      ng-model="$ctrl.depotId"
+      on-select="$ctrl.onSelect($item, $model)">
+      <ui-select-match placeholder="{{ 'FORM.SELECT.DEPOT' | translate }}">{{$select.selected.text}}</span></ui-select-match>
```
There is an unmatched `</span>` here.

jniles commented on client/src/modules/templates/bhInventorySelect.tmpl.html:
```diff
@@ -0,0 +1,24 @@
+    <ui-select name="inventory_id" 
+      ng-model="$ctrl.inventoryId"
+      on-select="$ctrl.onSelect($item, $model)">
+      <ui-select-match placeholder="{{ 'FORM.SELECT.INVENTORY' | translate }}">{{$select.selected.label}}</span></ui-select-match>
```
There is an unmatched `</span>` here.

jniles commented on client/src/modules/stock/lots/registry.js:
```diff
@@ -107,53 +134,91 @@ function StockLotsController(Stock, Notify,
 
-    vm.loading = true;
+  /**
+   * @function toggleLoadingIndicator
```
Nice documentation :+1:

jniles commented on client/src/modules/stock/lots/registry.js:
```diff
@@ -14,72 +15,98 @@ StockLotsController.$inject = [
   vm.filters = { lang: Languages.key };
   vm.formatedFilters = [];
+  vm.gridApi = {};
+  vm.gridOptions = {};
```
Why do you need to define `gridOptions` here?

jniles commented on client/src/modules/stock/lots/registry.js:
```diff
@@ -14,72 +15,98 @@ StockLotsController.$inject = [
+  var cacheKey = 'lot-grid';
+  var gridColumns;
+  var state;
+  var FILTER_BAR_HEIGHT = bhConstants.grid.FILTER_BAR_HEIGHT;
```
Where do you use this variable?

jniles commented on client/src/modules/stock/lots/registry.html:
```diff
@@ -1,4 +1,4 @@
-<div class="flex-header static">
+<div class="flex-header">
```
:+1:

jniles commented on client/src/js/components/bhInventorySelect.js:
```diff
@@ -0,0 +1,45 @@
+      disable          : '<?',
+      onSelectCallback : '&',
+      name             : '@?',
+      required         : '<?',      
```
Is `required` ever used?

jniles commented on client/src/js/components/bhDepotSelect.js:
```diff
@@ -0,0 +1,45 @@
+      disable          : '<?',
+      onSelectCallback : '&',
+      name             : '@?',
+      required         : '<?',      
```
Is `required` ever used?

jniles reviewed (commented):
@dedrickenc, this code has been reviewed.

The logic looks correct, and the code is relatively clean.  Good job!

Almost all of the comments deal with maintainability of the code.  There are a few places were there is broken HTML, or variables that are defined and never used, or parameters that are inconsistently named.  Those will make maintaining the code really difficult.  Since they do not affect the logic at all, its worth it to get them right the first time.

DedrickEnc committed:
fix(test): Update test

Update the end to end test to reflect the change made on our depot select inventory select
components and inventory select

DedrickEnc committed:
test(test): update test e2e

jniles commented on client/src/modules/templates/bhInventorySelect.tmpl.html:
```diff
@@ -0,0 +1,27 @@
+    <ng-transclude></ng-transclude>
+    <ui-select 
+      name="inventory_uuid" 
+      ng-model="$ctrl.inventoryUuid"
```
Much better :+1:

jniles commented on client/src/modules/stock/lots/registry.js:
```diff
@@ -107,53 +131,91 @@ function StockLotsController(Stock, Notify,
+
+        return load(Stock.lotFilters.formatHTTP(true));
+      })
+      .catch(angular.noop);
```
This should probably be handled... but I don't think it will ever get called since `load()` has its own error handling.

jniles reviewed (approved):
Everything looks addressed. Well done!
